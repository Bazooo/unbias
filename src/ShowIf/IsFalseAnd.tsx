import { ShowIfElement } from './ShowIf'

interface IsFalseAndProps extends ShowIfElement {
  condition: Exclude<ShowIfElement['condition'], undefined>
}

const IsFalseAnd = ({ children }: IsFalseAndProps) => {
  /* istanbul ignore next */
  const {} = 'IS_FALSE_AND'
  return children
}

export default IsFalseAnd
