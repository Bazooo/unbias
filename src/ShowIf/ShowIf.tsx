import React from 'react'

type ShowIfCondition = boolean | (() => boolean)

export interface ShowIfElementProps {
  condition?: ShowIfCondition
  children: JSX.Element
}

interface ShowIfElementType {
  (): any
  showElementType: 'IS_TRUE' | 'IS_FALSE_AND' | 'OR_ELSE'
}

type ShowIfElement = React.ReactElement<ShowIfElementProps, ShowIfElementType>

interface ShowIfProps {
  condition: ShowIfCondition
  children: ShowIfElement | ShowIfElement[]
}

const ShowIf = ({ condition, children }: ShowIfProps) => {
  let trueChild: React.ReactElement<ShowIfElementProps> | null = null
  let elseChild: React.ReactElement<ShowIfElementProps> | null = null
  const falseChildren: React.ReactElement<ShowIfElementProps>[] = []

  // The rendered children
  let renderedChildren: JSX.Element | null = null

  if (Array.isArray(children)) {
    children.forEach((child, i) => {
      switch (child.type.showElementType) {
        case 'IS_TRUE':
          trueChild = child
          break
        case 'IS_FALSE_AND':
          falseChildren.push(child)
          break
        case 'OR_ELSE':
          elseChild = child
          break
        default:
          throw new Error(`ShowIf child[${i}] is invalid!`)
      }
    })
  } else {
    switch (children.type.showElementType) {
      case 'IS_TRUE':
        trueChild = children
        break
      case 'IS_FALSE_AND':
        falseChildren.push(children)
        break
      case 'OR_ELSE':
        elseChild = children
        break
      default:
        throw new Error(`ShowIf child is invalid!`)
    }
  }

  if (conditionCheck(condition)) {
    renderedChildren = trueChild && trueChild.props.children
  } else {
    const goodFalseChildren = falseChildren.filter(falseChild => conditionCheck(falseChild.props.condition)).map(falseChild => falseChild.props.children)

    if (goodFalseChildren.length > 0) {
      renderedChildren = <>{goodFalseChildren}</>
    } else {
      renderedChildren = elseChild && elseChild.props.children
    }
  }

  return <>{renderedChildren}</>
}

const conditionCheck = (condition?: ShowIfCondition) => {
  if (typeof condition === 'function') {
    return condition()
  } else {
    return condition
  }
}

export default ShowIf
