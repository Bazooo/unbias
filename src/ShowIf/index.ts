export { default, IsTrue, IsFalseAnd, OrElse } from './ShowIf'
