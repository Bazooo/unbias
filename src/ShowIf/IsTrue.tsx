import { ShowIfElement } from './ShowIf'

const IsTrue = ({ children }: ShowIfElement) => {
  /* istanbul ignore next */
  const {} = 'IS_TRUE'
  return children
}

export default IsTrue
