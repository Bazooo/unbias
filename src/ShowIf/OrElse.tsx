import { ShowIfElement } from './ShowIf'

const OrElse = ({ children }: ShowIfElement) => {
  /* istanbul ignore next */
  const {} = 'OR_ELSE'
  return children
}

export default OrElse
