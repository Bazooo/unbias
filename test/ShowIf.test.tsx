import React from 'react'
import { shallow } from 'enzyme'

import ShowIf, { IsTrue, OrElse, IsFalseAnd } from '../src/ShowIf'

describe('IsTrue', () => {
  it('should return children', () => {
    const wrapper = shallow(
      <IsTrue><span>test</span></IsTrue>
    )

    expect(wrapper.find('span').exists()).toBeTruthy()
    expect(wrapper).toMatchSnapshot()
  })

  it('should show if condition is truthy', () => {
    const wrapper = shallow(
      <ShowIf condition={true}>
        <IsTrue><span id='show'>works</span></IsTrue>
      </ShowIf>
    )

    expect(wrapper.find('span#show').exists()).toBeTruthy()
    expect(wrapper).toMatchSnapshot()
  })

  it('should show nothing if condition is falsy', () => {
    const wrapper = shallow(
      <ShowIf condition={false}>
        <IsTrue><span id='show'>works</span></IsTrue>
      </ShowIf>
    )

    expect(wrapper.find('span#show').exists()).toBeFalsy()
    expect(wrapper).toMatchSnapshot()
  })
})

describe('IsFalseAnd', () => {
  it('should return children', () => {
    const wrapper = shallow(
      <IsFalseAnd condition={true}><span>test</span></IsFalseAnd>
    )

    expect(wrapper.find('span').exists()).toBeTruthy()
    expect(wrapper).toMatchSnapshot()
  })
  it('should show if condition is falsy (F)[T]', () => {
    const wrapper = shallow(
      <ShowIf condition={false}>
        <IsFalseAnd condition={true}><span id='show'>works</span></IsFalseAnd>
      </ShowIf>
    )

    expect(wrapper.find('span#show').exists()).toBeTruthy()
    expect(wrapper).toMatchSnapshot()
  })

  it('should show nothing if condition is falsy but component condition is false (F)[F]', () => {
    const wrapper = shallow(
      <ShowIf condition={false}>
        <IsFalseAnd condition={false}><span id='show'>works</span></IsFalseAnd>
      </ShowIf>
    )

    expect(wrapper.find('span#show').exists()).toBeFalsy()
    expect(wrapper).toMatchSnapshot()
  })

  it('should show nothing if condition is truthy (T)[T,F]', () => {
    const wrapper = shallow(
      <ShowIf condition={true}>
        <IsFalseAnd condition={true}><span id='show'>works</span></IsFalseAnd>
        <IsFalseAnd condition={false}><span id='show2'>works</span></IsFalseAnd>
      </ShowIf>
    )

    expect(wrapper.find('span#show').exists()).toBeFalsy()
    expect(wrapper.find('span#show2').exists()).toBeFalsy()
    expect(wrapper).toMatchSnapshot()
  })

  it('should show only true conditions and in order (F)[T,F,T,F,T]', () => {
    const wrapper = shallow(
      <ShowIf condition={false}>
        <IsFalseAnd condition={true}><span id='show'>works</span></IsFalseAnd>
        <IsFalseAnd condition={false}><span id='show2'>works</span></IsFalseAnd>
        <IsFalseAnd condition={true}><span id='show3'>works</span></IsFalseAnd>
        <IsFalseAnd condition={false}><span id='show4'>works</span></IsFalseAnd>
        <IsFalseAnd condition={true}><span id='show5'>works</span></IsFalseAnd>
      </ShowIf>
    )

    // test existance
    expect(wrapper.find('span#show').exists()).toBeTruthy()
    expect(wrapper.find('span#show2').exists()).toBeFalsy()
    expect(wrapper.find('span#show3').exists()).toBeTruthy()
    expect(wrapper.find('span#show4').exists()).toBeFalsy()
    expect(wrapper.find('span#show5').exists()).toBeTruthy()

    // test ordering
    expect(wrapper.find('span').first().prop('id') === 'show').toBeTruthy()
    expect(wrapper.find('span').at(1).prop('id') === 'show3').toBeTruthy()
    expect(wrapper.find('span').at(2).prop('id') === 'show5').toBeTruthy()

    expect(wrapper).toMatchSnapshot()
  })

  it('should accept functions as condition', () => {
    const truthyFunc = () => true
    const falsyFunc = () => false
    const wrapper = shallow(
      <ShowIf condition={false}>
        <IsFalseAnd condition={truthyFunc}><span id='show'>works</span></IsFalseAnd>
        <IsFalseAnd condition={falsyFunc}><span id='show2'>works</span></IsFalseAnd>
      </ShowIf>
    )

    expect(wrapper.find('span#show').exists()).toBeTruthy()
    expect(wrapper.find('span#show2').exists()).toBeFalsy()
    expect(wrapper).toMatchSnapshot()
  })
})

describe('OrElse', () => {
  it('should return children', () => {
    const wrapper = shallow(
      <OrElse><span>test</span></OrElse>
    )

    expect(wrapper.find('span').exists()).toBeTruthy()
    expect(wrapper).toMatchSnapshot()
  })

  it('should show if condition is falsy', () => {
    const wrapper = shallow(
      <ShowIf condition={false}>
        <OrElse><span id='show'>works</span></OrElse>
      </ShowIf>
    )

    expect(wrapper.find('span#show').exists()).toBeTruthy()
    expect(wrapper).toMatchSnapshot()
  })

  it('should show nothing if condition is truthy', () => {
    const wrapper = shallow(
      <ShowIf condition={true}>
        <OrElse><span id='show'>works</span></OrElse>
      </ShowIf>
    )

    expect(wrapper.find('span#show').exists()).toBeFalsy()
    expect(wrapper).toMatchSnapshot()
  })
})

describe('ShowIf', () => {
  it('should accept functions as condition', () => {
    const truthyFunc = () => true
    const falsyFunc = () => false
    const wrapper = shallow(
      <div>
        <ShowIf condition={truthyFunc}>
          <IsTrue><span id='show'>works</span></IsTrue>
        </ShowIf>
        <ShowIf condition={falsyFunc}>
          <OrElse><span id='show2'>works</span></OrElse>
        </ShowIf>
      </div>
    )

    expect(wrapper.find('span#show').exists()).toBeTruthy()
    expect(wrapper.find('span#show2').exists()).toBeTruthy()
    expect(wrapper).toMatchSnapshot()
  })

  it('works', () => {
    const wrapper = shallow(
      <ShowIf condition={true}>
        <IsTrue><span id='show'>works</span></IsTrue>
        <IsFalseAnd condition={true}><span id='show2'>works</span></IsFalseAnd>
        <IsFalseAnd condition={false}><span id='show3'>works</span></IsFalseAnd>
        <OrElse><span id='show4'>works</span></OrElse>
      </ShowIf>
    )

    expect(wrapper).toMatchSnapshot()

    expect(wrapper.find('span#show').exists()).toBeTruthy()
    expect(wrapper.find('span#show2').exists()).toBeFalsy()
    expect(wrapper.find('span#show3').exists()).toBeFalsy()
    expect(wrapper.find('span#show4').exists()).toBeFalsy()

    wrapper.setProps({ condition: false })

    expect(wrapper.find('span#show').exists()).toBeFalsy()
    expect(wrapper.find('span#show2').exists()).toBeTruthy()
    expect(wrapper.find('span#show3').exists()).toBeFalsy()
    expect(wrapper.find('span#show4').exists()).toBeFalsy()

    const firstChange = [
      (<IsTrue><span id='show'>works</span></IsTrue>),
      (<IsFalseAnd condition={true}><span id='show2'>works</span></IsFalseAnd>),
      (<IsFalseAnd condition={true}><span id='show3'>works</span></IsFalseAnd>),
      (<OrElse><span id='show4'>works</span></OrElse>)
    ]

    wrapper.setProps({ children: firstChange })

    expect(wrapper.find('span#show').exists()).toBeFalsy()
    expect(wrapper.find('span#show2').exists()).toBeTruthy()
    expect(wrapper.find('span#show3').exists()).toBeTruthy()
    expect(wrapper.find('span#show4').exists()).toBeFalsy()

    const secondChange = [
      (<IsTrue><span id='show'>works</span></IsTrue>),
      (<IsFalseAnd condition={false}><span id='show2'>works</span></IsFalseAnd>),
      (<IsFalseAnd condition={false}><span id='show3'>works</span></IsFalseAnd>),
      (<OrElse><span id='show4'>works</span></OrElse>)
    ]

    wrapper.setProps({ children: secondChange })

    expect(wrapper.find('span#show').exists()).toBeFalsy()
    expect(wrapper.find('span#show2').exists()).toBeFalsy()
    expect(wrapper.find('span#show3').exists()).toBeFalsy()
    expect(wrapper.find('span#show4').exists()).toBeTruthy()
  })

  it('should throw an error if wrong child', () => {
    const wrapper = () => shallow(
      <ShowIf condition={true}>
        <span>wrong</span>
      </ShowIf>
    )

    expect(wrapper).toThrowError('ShowIf child is invalid!')
  })

  it('should throw an error if wrong children', () => {
    const wrapper = () => shallow(
      <ShowIf condition={true}>
        <IsTrue><span>test</span></IsTrue>
        <span>wrong</span>
      </ShowIf>
    )

    expect(wrapper).toThrowError('ShowIf child[1] is invalid!')
  })
})
