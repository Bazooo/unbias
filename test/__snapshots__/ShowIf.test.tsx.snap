// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IsFalseAnd should accept functions as condition 1`] = `
<Fragment>
  <span
    id="show"
  >
    works
  </span>
</Fragment>
`;

exports[`IsFalseAnd should return children 1`] = `
<span>
  test
</span>
`;

exports[`IsFalseAnd should show if condition is falsy (F)[T] 1`] = `
<Fragment>
  <span
    id="show"
  >
    works
  </span>
</Fragment>
`;

exports[`IsFalseAnd should show nothing if condition is falsy but component condition is false (F)[F] 1`] = `<Fragment />`;

exports[`IsFalseAnd should show nothing if condition is truthy (T)[T,F] 1`] = `<Fragment />`;

exports[`IsFalseAnd should show only true conditions and in order (F)[T,F,T,F,T] 1`] = `
<Fragment>
  <span
    id="show"
  >
    works
  </span>
  <span
    id="show3"
  >
    works
  </span>
  <span
    id="show5"
  >
    works
  </span>
</Fragment>
`;

exports[`IsTrue should return children 1`] = `
<span>
  test
</span>
`;

exports[`IsTrue should show if condition is truthy 1`] = `
<Fragment>
  <span
    id="show"
  >
    works
  </span>
</Fragment>
`;

exports[`IsTrue should show nothing if condition is falsy 1`] = `<Fragment />`;

exports[`OrElse should return children 1`] = `
<span>
  test
</span>
`;

exports[`OrElse should show if condition is falsy 1`] = `
<Fragment>
  <span
    id="show"
  >
    works
  </span>
</Fragment>
`;

exports[`OrElse should show nothing if condition is truthy 1`] = `<Fragment />`;

exports[`ShowIf should accept functions as condition 1`] = `
<div>
  <ShowIf
    condition={[Function]}
  >
    <IsTrue>
      <span
        id="show"
      >
        works
      </span>
    </IsTrue>
  </ShowIf>
  <ShowIf
    condition={[Function]}
  >
    <OrElse>
      <span
        id="show2"
      >
        works
      </span>
    </OrElse>
  </ShowIf>
</div>
`;
